File: src/linked_list.py
"""Contains class function for generating linked lists."""


class LinkedList(object):
    """Creating LinkedList class."""

    def __init__(self, iterable=()):
        """Initialize LinkedList class."""
        self.head = None
        self._counter = 0
        if isinstance(iterable, (str, list, tuple)):
            for item in iterable:
                self.push(item)

    def push(self, val):
        """Create push method of LinkedList."""
        self.head = Node(val, self.head)
        self._counter += 1

    def pop(self):
        """Remove and returns value of head node."""
        if self.head is None:
            raise IndexError('List is empty.')
        output = self.head.data
        self.head = self.head.next
        self._counter -= 1
        return output

    def size(self):
        """Return length of linked list."""
        return self._counter

    def __len__(self):
        """Return length of linked list."""
        return self._counter

    def search(self, val):
        """Return node containing given value."""
        current_node = self.head
        while current_node:
            if current_node.data == val:
                return current_node
            current_node = current_node.next

    def remove(self, val):
        """Remove given node from linked list."""
        current_node = self.head
        while current_node.next is not None:
            if current_node.data == val:
                current_node.pop()
            current_node = current_node.next


class Node(object):
    """Creating Node class."""
    def __init__(self, data, next):
        self.data = data
        self.next = next

File: js/linked-list.js
'use strict';

class LinkedList {
  constructor(iterable=[]) {
    this.head = null;
    this._counter = 0;
    if(iterable.length > 0){
      iterable.forEach(x => this.push(x));
    }
  }

  push(val) {
    this.head = new Node(val, this.head);
    this._counter ++;
  }

  pop() {
    if(this.head === null) {
      throw 'Cannot pop from empty list.';
    }
    let output = this.head.data;
    this.head = this.head.next;
    this._counter --;
    return output;
  }

  size() {
    return this._counter;
  }

  search(val) {
    let currentNode = this.head;
    while (currentNode !== null) {
      if (currentNode.data === val) {
        return currentNode;
      }
      currentNode = currentNode.next;
    }
    throw 'No node in LL containing val.';
  }

  remove(val){
    if (this.size() === 0) {
      throw 'Cannot use remove on empty List.'
    }
    if (this.head.data === val) {
      this.head = this.head.next;
      this._counter --;
      return;
    }
    let currentNode = this.head.next;
    let prevNode = this.head;
    let nodeRemoved = false;
    while (currentNode !== null && nodeRemoved === false) {
      if (currentNode.data === val) {
        prevNode.next = currentNode.next;
        currentNode.next = null;
        nodeRemoved = true;
        break;
      }
      prevNode = currentNode;
      currentNode = currentNode.next;
    }
    if (nodeRemoved === false) {
      throw 'Value given not in Linked List.';
    }
  }

  display(){
    let currentNode = this.head;
    let listDisplay = [];
    while (currentNode !== null) {
      listDisplay.push(currentNode.data);
      currentNode = currentNode.next;
    }
    return listDisplay.join(', ');
  }
}

class Node {
  constructor(data, next=null){
    this.data = data
    this.next = next
  }
}

module.exports = {LinkedList, Node};
if(require.main === module){
  var linkedList = new LinkedList([1, 2, 3, 4]);
  console.log(linkedList.display());
  console.log(linkedList.size());
  console.log(linkedList.remove(6));
}
